### PR TITLE
wp-env: override generated file directory with environment variable

### PR DESCRIPTION
**Pull request by noahtallen**
## Description
- Adds a non-hidden default directory for Linux (`~/wp-env`) 
- Allows a user to override the directory in which wp-env creates generated files with the `WP_ENV_HOME` environment variable.

Work towards #20180, but will need to continue solving it in a follow-up.

## How has this been tested?
I used the environment variable locally and verified generated files were placed there. I also checked that the default location is still used if the environment variable is not specified.

I also tested that, if the container is running, the start command works successfully after deleting the wp-env home directory.

## Types of changes
New Feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/env/CHANGELOG.md
Comment (epiqueras) [inline packages/env/CHANGELOG.md]:
```suggestion
- You may now override the directory in which `wp-env` creates generated files with the `WP_ENV_SOURCE` environment variable. (The default directory is `~/.wp-env/`.)
```

Shouldn't this also be on top since it came after the other feature?

File: packages/env/README.md
Comment (epiqueras) [inline packages/env/README.md]:
```suggestion
By default, the files `wp-env` generates are located at `~/.wp-env/$md5_of_project_path`. To change this location, set the `WP_ENV_SOURCE` environment variable. This changes the `~/.wp-env/` part of the path to wherever you wish. For example, running `WP_ENV_SOURCE="something" wp-env start` will download the project files to the directory `./something/$md5_of_project_path` (relative to the current directory).
```

File: packages/env/lib/config.js
Comment (epiqueras) [inline packages/env/lib/config.js]: It's probably not needed, but I'd sleep better if we use `path.join` instead of hardcoding a path separator in a template literal like that.

**Review by epiqueras — CHANGES_REQUESTED** [review]

**Comment by noisysocks** [comment]
In the code for `wp-env` we call the `~/.wp-env` directory the "work directory". I'm unsure if that's the best name (I struggled with choosing one!), but nonetheless thinking that we should be consistent and name this new variable something like `WP_ENV_WORK_DIR`.

File: packages/env/lib/config.js
Comment (noisysocks) [reply packages/env/lib/config.js]: Yes we'll need to use `path.join()` so that this works on Windows.

**Comment by ockham** [comment]
Thanks Noah! This allows me to run `wp-env` at least once! :sweat_smile: 

```bash
~/src/gutenberg$ WP_ENV_SOURCE=~/wpenv npx wp-env start
> ✔ WordPress started. (in 296s 268ms)
~/src/gutenberg$ WP_ENV_SOURCE=~/wpenv npx wp-env stop
✔ Stopped WordPress. (in 4s 979ms)
> ~/src/gutenberg$ WP_ENV_SOURCE=~/wpenv npx wp-env start
✖ EACCES: permission denied, open '~/wpenv/6ac6b03bd54e5e922f0d180f6d7ff25d/tests-WordPress/wp-config.php'
[Error: EACCES: permission denied, open '~/wpenv/6ac6b03bd54e5e922f0d180f6d7ff25d/tests-WordPress/wp-config.php'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: '~/wpenv/6ac6b03bd54e5e922f0d180f6d7ff25d/tests-WordPress/wp-config.php'
}
```

FWIW:
```bash
-rw-r--r-- 1 www-data www-data 3187 Feb 17 14:19 ~/wpenv/6ac6b03bd54e5e922f0d180f6d7ff25d/tests-WordPress/wp-config.php
```

whereas the other files in that directory have my user and group as their owner.

**Comment by epiqueras** [comment]
I guess we need to set the owner and permissions of the file explicitly for Linux.

File: packages/env/CHANGELOG.md
Comment (noahtallen) [reply packages/env/CHANGELOG.md]: sure :)

File: packages/env/lib/config.js
Comment (noahtallen) [reply packages/env/lib/config.js]: Good call.

**Comment by noahtallen** [comment]
Per the permissions issue, it looks as if the basic stuff is working well. So I'm not sure it's a node filesystem issue since we use [nodegit](https://github.com/nodegit/nodegit) to download the sourcecode that's has this permissions issue. ([Here's where we use it](https://github.com/WordPress/gutenberg/blob/28c49aee97a9e9c9111891211b9666bd8ca0c001/packages/env/lib/download-source.js#L53-L60).)

Do we need to run chmod and/or chown on every downloaded file?

**Comment by noahtallen** [comment]
Actually, scratch that last comment.

- web servers on Ubuntu are under the www-data user. (https://askubuntu.com/questions/873839/what-is-the-www-data-user)
- wp-config.php is generated by the `wp core install` command

I think the install command, which is run by docker-compose, is just generating the file under that web server user.

So, should we chmod/chown on the wp-config file for any WordPress source?

**Comment by noahtallen** [comment]
Hm, maybe it's something else entirely. On this branch:

```sh
~/source/gutenberg(add/wp-env-support-for-source-override) » packages/env/bin/wp-env start
✖ Error while running docker-compose command.
Starting 31911d623e75f345e9ed328b9f48cff6_mysql_1 ... done
Starting 31911d623e75f345e9ed328b9f48cff6_wordpress_1 ... done
Error: 'wp-config.php' not found.
Either create one manually or use `wp config create`.
```

I get this on a fresh install (after deleting the wp-env directory for the project) on this branch even on macOS. lol. I've been testing in a separate plugin, which does not specify a GitHub source for WordPress, so it may just be that the Gutenberg config has been broken for a while... 🤔

**Comment by epiqueras** [comment]
I don't get that error.

File: packages/env/README.md
Comment (epiqueras) [inline packages/env/README.md]:
```suggestion
`wp-env` creates generated files in the `wp-env` home directory. By default, this is `~/.wp-env`. The exception is Linux, where files are placed at `~/wp-env` [for compatibility with Snap Packages](https://github.com/WordPress/gutenberg/issues/20180#issuecomment-587046325). The `wp-env` home directory contains a subdirectory for each project named `/$md5_of_project_path`. To change the `wp-env` home directory, set the `WP_ENV_HOME` environment variable. For example, running `WP_ENV_HOME="something" wp-env start` will download the project files to the directory `./something/$md5_of_project_path` (relative to the current directory).
```

File: packages/env/lib/config.js
Comment (epiqueras) [inline packages/env/lib/config.js]:
```suggestion
 * Gets the `wp-env` home directory in which generated files are created.
```

File: packages/env/lib/config.js
Comment (epiqueras) [inline packages/env/lib/config.js]:
```suggestion
 * @return {string} The absolute path to the `wp-env` home directory.
```

File: packages/env/lib/config.js
Comment (epiqueras) [inline packages/env/lib/config.js]: I think we can drop the `join`.

**Review by epiqueras — CHANGES_REQUESTED** [review]

**Comment by noahtallen** [comment]
The old command [relied on this script](https://github.com/WordPress/wordpress-develop/blob/master/tools/local-env/scripts/install.js) to properly initialize the config file. I'm a bit out of my depth here; @noisysocks do you have opinions about how to handle this?

**Comment by epiqueras** [comment]
Doesn't `wp core install` create it for you?

**Comment by noahtallen** [comment]
> Doesn't wp core install create it for you?

That's what I expected to happen as well, but if I run these commands on this branch, I consistently get this issue.

```zsh
~/source/gutenberg(add/wp-env-support-for-source-override) » rm -rf ~/.wp-env
~/source/gutenberg(add/wp-env-support-for-source-override) » packages/env/bin/wp-env start
✖ Error while running docker-compose command.
Starting 31911d623e75f345e9ed328b9f48cff6_mysql_1 ... done
Starting 31911d623e75f345e9ed328b9f48cff6_wordpress_1 ... done
Error: 'wp-config.php' not found.
Either create one manually or use `wp config create`.
```

**Comment by noahtallen** [comment]
For that matter, I also get the error on master, so we could resolve this in a follow up.

**Comment by epiqueras** [comment]
Add logs to figure out on what line that throws. It's probably related to the new order of commands.

**Comment by noahtallen** [comment]
Ah, apparently this condition happens when:
- you already have the docker instance started
- you delete the ~/.wp-env directory
- you run wp-env start again

It looks like it can be fixed by stopping the docker instance before running wp-env start.

I think we should handle this case gracefully, but it definitely does not appear to be the reported issue.

So we're back at https://github.com/WordPress/gutenberg/pull/20253#issuecomment-587183306, which I think means we just need to set the correct permissions on the wp-config file.

**Comment by epiqueras** [comment]
Yes, let's just set the permissions here to be safe but also `resetDatabase` at the start of `start` if the `wp-env` work directory is missing to simply avoid the error you encountered.

File: packages/env/lib/env.js
Comment (epiqueras) [inline packages/env/lib/env.js]:
```suggestion
 * Sets the correct user and permissions on the wp-config.php file.
```

File: packages/env/lib/env.js
Comment (epiqueras) [inline packages/env/lib/env.js]: Wouldn't `coreSourcePath` be more consistent?

**Review by epiqueras — CHANGES_REQUESTED** [review]

File: packages/env/lib/env.js
Comment (noahtallen) [reply packages/env/lib/env.js]: Sure. :)

**Comment by noahtallen** [comment]
@epiqueras Instead of resetting the database, I run `stop` at the beginning of running `start`. I think it solves the issue more cleanly than resetting the database, at least with the way I was thinking about trying to call the reset command under the right situation. We can always call stop at the beginning of the command without having to think about different situations.

- Database don't go away if something happens to the wp-env folder.
- You effectively "restart" the container by running the start command twice, which might be a nice solution.
- Fixes the issue with running configuration while the instance is still up.
- Don't have to worry about detecting cases where the wp-env home folder doesn't exist, since the stop command creates that folder for us.

**Comment by noahtallen** [comment]
@ockham I think this should be ready to test again; can you let me know if running `./packages/env/bin/wp-env start` on this branch from the Gutenberg root works on Linux?

**Comment by epiqueras** [comment]
@noahtallen I don't see how running `stop` would fix the issue.

If the databases stay, `wp core install` won't work and won't generate a new `wp-config.php`

**Comment by noisysocks** [comment]
`wp core install` doesn't generate a `wp-config.php`. This file is created and/or updated when the `wordpress` Docker container starts and runs its `docker-entrypoint.sh` script. 

We should test if generating a `wp-config.php` ourselves using `wp config generate` causes the config file to have correct permissions. If so, then let's do that. If not, we may need to look at creating a `wp-config.php` using Node (instead of WP CLI) or using `chmod`.

**Comment by epiqueras** [comment]
I see, that's why `stop` => `start` works.

>We should test if generating a wp-config.php ourselves using wp config generate causes the config file to have correct permissions. If so, then let's do that. If not, we may need to look at creating a wp-config.php using Node (instead of WP CLI) or using chmod.

That was already solved in this PR by using `chown`.

File: packages/env/lib/config.js
Comment (noisysocks) [inline packages/env/lib/config.js]: I'd suggest renaming to `getHomeDirectory()`. `WpEnv` is redundant because all of this code refers to `wp-env`, and it's best to avoid abbreviations where we can.

File: packages/env/lib/env.js
Comment (noisysocks) [inline packages/env/lib/env.js]: Why is this necessary? An inline comment might be helpful for future readers.

We'll also need to `await` this as `stop()` is an async function.

File: packages/env/lib/env.js
Comment (noisysocks) [inline packages/env/lib/env.js]:
```suggestion

	// Get the UID and GID of an existing file that works correctly.
```

**Review by noisysocks — CHANGES_REQUESTED** [review]
I haven't tested this but the changes make sense. My only blocking comment is that we need to `await` the call to `stop()`.

**Comment by noahtallen** [comment]
Thanks for the tips! Those changes should be implemented now.

> This file is created and/or updated when the wordpress Docker container starts and runs its docker-entrypoint.sh script.

Ahhhh that explains a lot. Very helpful. Now that I know that, I've moved the `chown` command to be immediately after dockerCompose.upMany so that we have the right permissions for the wp install and other commands. (I'm not really sure where the permissions error happens).

File: packages/env/lib/env.js
Comment (epiqueras) [inline packages/env/lib/env.js]:
```suggestion
		 * If the Docker image is already running and the `wp-env` files have been
```

File: packages/env/lib/env.js
Comment (epiqueras) [inline packages/env/lib/env.js]:
```suggestion
		 * the container before continuing allows the docker entrypoint script, which restores the files,
```

**Review by epiqueras — APPROVED** [review]

File: packages/env/lib/env.js
Comment (noahtallen) [reply packages/env/lib/env.js]: Jeez, you'd think I'd remember

**Comment by sainthkh** [comment]
When I run it without `sudo` command, I got this error:

```
✖ Error while running docker-compose command.
Couldn't connect to Docker daemon at http+docker://localhost - is it running?

If it's at a non-standard location, specify the URL with the DOCKER_HOST environment variable.
```

If I use `sudo` command, I've got this error:

```
✖ Error while running docker-compose command.
Starting 841be7b26bd583f294009fefb90de98d_mysql_1 ... done
Starting 841be7b26bd583f294009fefb90de98d_wordpress_1 ... done
mysqlcheck: Got error: 1049: Unknown database 'wordpress' when selecting the database
```

Do you have any idea why these are happening?

**Comment by epiqueras** [comment]
Update Docker Desktop.

**Comment by sainthkh** [comment]
Reinstalled docker-ce and docker-compose. After `sudo ./packages/env/bin/wp-env start`, I got this error:

```
✖ Error while running docker-compose command.
Starting 841be7b26bd583f294009fefb90de98d_mysql_1 ... done
Starting 841be7b26bd583f294009fefb90de98d_wordpress_1 ... done
mysqlcheck: Got error: 2002: Can't connect to MySQL server on 'mysql' (115) when trying to connect
```

My env:

```
me@me:~/wp/gutenberg$ docker --version
Docker version 19.03.6, build 369ce74a3c
me@me:~/wp/gutenberg$ docker-compose --version
docker-compose version 1.25.4, build 8d51620a
```

**Comment by epiqueras** [comment]
>Reinstalled docker-ce and docker-compose

How are you doing that? Docker comes in a desktop bundle now.

**Comment by sainthkh** [comment]
I followed an instruction in this doc: https://docs.docker.com/install/linux/docker-ce/ubuntu/

**Comment by epiqueras** [comment]
Any ideas @ockham?

**Comment by ockham** [comment]
Just retested this (after deleting my previous `wp-env` directory). Unfortunately, now the first attempt is also failing:

```bash
$ npx wp-env start
✖ EPERM: operation not permitted, chown '/home/me/wp-env/6ac6b03bd54e5e922f0d180f6d7ff25d/WordPress/wp-config.php'
[Error: EPERM: operation not permitted, chown '/home/me/wp-env/6ac6b03bd54e5e922f0d180f6d7ff25d/WordPress/wp-config.php'] {
  errno: -1,
  code: 'EPERM',
  syscall: 'chown',
  path: '/home/me/wp-env/6ac6b03bd54e5e922f0d180f6d7ff25d/WordPress/wp-config.php'
}
```

Looks like the [`chown`](https://github.com/WordPress/gutenberg/pull/20253/files#diff-ca5826745aaf2e634a31b5f8df236d96R410), while a good idea, isn't working :slightly_frowning_face: 

Let's go back to the state of this PR before the `chown` I guess? Working at least once is better than not working :sweat_smile:

**Comment by ockham** [comment]
> Any ideas @ockham?

Just saw this -- unfortunately, I don't have any ideas; I'm seeing different errors.

@sainthkh I'm using the 'Snap' package for Docker that Ubuntu provides (in its Software Center) -- have you tried that? (You'd probably need to remoive your current Docker setup :sweat_smile: )

**Review by ockham — CHANGES_REQUESTED** [review]
Blocking merge b/c of https://github.com/WordPress/gutenberg/pull/20253#issuecomment-587498759

**Comment by ockham** [comment]
As a follow-up to this PR, we might consider generating the `wp-config.php` as suggested by @noisysocks here (https://github.com/WordPress/gutenberg/pull/20253#issuecomment-587249535):

> We should test if generating a `wp-config.php` ourselves using `wp config generate` causes the config file to have correct permissions.

(Maybe this stands a better chance of success since it's called from within the container. Bit of a hand-waving argument here :sweat_smile: )

Suggesting to do this as a follow-up, since all of this is a massive can of worms, and I think it makes sense to divide and conquer: Make this PR only about the env variable (and its Linux default), and tackle `wp-config` permissions separately.

**Comment by epiqueras** [comment]
>Suggesting to do this as a follow-up, since all of this is a massive can of worms, and I think it makes sense to divide and conquer: Make this PR only about the env variable (and its Linux default), and tackle wp-config permissions separately.

Agreed, unless we can fix `chown` somehow.

**Comment by ockham** [comment]
> > Suggesting to do this as a follow-up, since all of this is a massive can of worms, and I think it makes sense to divide and conquer: Make this PR only about the env variable (and its Linux default), and tackle wp-config permissions separately.
> 
> Agreed, unless we can fix `chown` somehow.

My gut says that the chown issue is actually the same as the previous one with accessing the `wp-config`: When done from outside the container, we plainly lack permissions, no matter what we try to do with that file. (But yeah, this is clearly not my domain of expertise, so I might be wrong.)

**Comment by epiqueras** [comment]
It makes sense, you need permissions to change permissions. We would need to `sudo` somehow.

**Comment by noahtallen** [comment]
Sad to hear that it did not work. I've removed chown from this branch now; I agree that it makes sense to split this up and continue trying to fix that issue after merging the other stuff here :)

**Comment by noahtallen** [comment]
> mysqlcheck: Got error: 2002: Can't connect to MySQL server on 'mysql' (115) when trying to connect

@sainthkh Thanks for the test! Interestingly, I also saw a similar error yesterday when running `docker-compose run cli wp config create`.

**Comment by noahtallen** [comment]
cc @ockham and @noisysocks for ✅

**Review by noisysocks — APPROVED** [review]
Looks good! :shipit:

**Comment by sainthkh** [comment]
I changed to snap docker. But there was no difference. 

```
✖ Error while running docker-compose command.
Starting 841be7b26bd583f294009fefb90de98d_mysql_1 ... done
Starting 841be7b26bd583f294009fefb90de98d_wordpress_1 ... done
mysqlcheck: Got error: 1049: Unknown database 'wordpress' when selecting the database
```

I either get `no 'wordpress'` or `cannot connect` error.

**Review by ockham — APPROVED** [review]
Thanks Noah! We're back to [this behavior](https://github.com/WordPress/gutenberg/pull/20253#issuecomment-586871441) -- which, as discussed, is an improvement over the previous state of things!

**Comment by noahtallen** [comment]
@ockham & @sainthkh Thanks for the continued testing. I can keep working towards https://github.com/WordPress/gutenberg/issues/20180, but it is a bit tough since I don't have a way to test that the changes work. :) The two things I'm thinking of for next steps are:
1. Try generating the wp-config.php file ourselves
2. Look up mysql connection bugs. I wonder if we need to define a network for the different Docker services?

**Comment by epiqueras** [comment]
Has anyone reproduced the connection issue? It might be a third party problem.

**Comment by noahtallen** [comment]
I did see a similar error locally when I was playing around with generating the config file, so I'll keep an eye out for it

**Comment by gglnx** [comment]
```bash
Starting 8dd17dfdf919942794ccb8bf7b39ce80_mysql_1 ... done
Starting 8dd17dfdf919942794ccb8bf7b39ce80_wordpress_1 ... done
mysqlcheck: Got error: 1049: Unknown database 'wordpress' when selecting the database

Starting 8dd17dfdf919942794ccb8bf7b39ce80_mysql_1 ... done
Starting 8dd17dfdf919942794ccb8bf7b39ce80_wordpress_1 ... done
mysqlcheck: Got error: 1049: Unknown database 'wordpress' when selecting the database
```

I've got the same problem. Tried it with `sudo` and without it. Using the latest Docker Desktop (Mac, 2.2.0.3).

**Comment by jorgefilipecosta** [comment]
Hi @noahtallen should this PR be backported into WordPress 5.4? If yes, feel free to add the label backport to wp core.

**Comment by noahtallen** [comment]
> Hi @noahtallen should this PR be backported into WordPress 5.4?

If the other PRs for wp-env are being backported, I don't see why not :)

**Pull request merged**
